Firestore Field Path

chemelnucfin opened the pull request:
@coryan I wrote this class for google-cloud-python and I wanted to write some C++.  I don't recall a lot of the details with RVO and copy elision, but maybe it's a start somewhere.  No urgency, just a toy program.

coryan commented:
@chemelnucfin welcome!  Before we can incorporate your changes we need a few things:

* Can you describe what your objective is here?  Beyond this class, what do you want to achieve long term?  Support the Firestore API in full?  A few helper classes for it?  Maybe open a bug to describe the intent?
* You should include a CMake file for your classes, and it should be called from the top-level CMake.
  * Once you do that the builds will fail because your code does not follow our code formatting rules, the [CONTRIBUTING](CONTRIBUTING.md) file describes how to fix that.
* You should include a README.md file describing what the contents of the new firestore/ directory are.

If you do want to make a contribution around firestore I would be happy to review PRs, but I won't be able to do that except when my other work is done, so please be patient.

Thanks!

@mbrukman , @ghaisa FYI.

mbrukman commented:
/cc: @dmcgrath (FYI)

chemelnucfin commented:
@coryan Hi!  I'm not sure of my long term objective at this time.  I just figured if you guys need firestore C++ code, someone would probably write this eventually.  I was just translating this for fun.

I ran the clang-format on the cc and h files, but I'm not too sure how to set up the CMake files especially in relation to the overall client and testing.  

Just ping me if/when you have some free time.  Feel free to close this in the mean time.  I know you are working on bigtable and is a higher priority.

Thanks!

mbrukman commented:
@chemelnucfin — please also add an Apache 2.0 license header to each of your `*.h` and `*.cc` files; you can find such a header in any file of this repo. Your `CMakeLists.txt` file has that header.

chemelnucfin commented:
@mbrukman Hi, I added the apache header, but I'm not too sure of how to take care of the cmake stuff at this time.  Since it seems to also have to deal with a lot of other things.  I can look into it further later if desired.

I also moved the files to `google/client/` as I believe that is a better place to put it looking at the bigtable files.

Please let me know if I can be of help further.  Thanks.

coryan commented on firestore/client/field_path.cc:
```diff
@@ -0,0 +1,97 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "field_path.h"
```
This should be `#include "firestore/client/field_path.h"`.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 Google Inc.
```
s/Inc./LLC/

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLECLOUDCPP_FIRESTORE_GOOGLE_FIRESTORE_FIELD_PATH_H_
```
s/GOOGLECLOUDCPP/GOOGLE_CLOUD_CPP/

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+  static std::regex simple_field_name;
+
+  // Construct a new FieldPath by appending a field path string
+  const FieldPath append(const std::string& path) const;
```
Returning a `const` has no effect, they can be assigned to non-const objects.  Please remove here and elsewhere.

chemelnucfin replied:
@coryan I have addressed your other comments.  I recalled that there was a difference returning const objects, which I have found through [here](https://www.linuxtopia.org/online_books/programming_books/thinking_in_c++/Chapter08_014.html).

Perhaps things have changed or that it is irrelevant in this case?

coryan replied:
Thanks!  I learned something today.  I remain ~stubborn~ unconvinced.  Your type is closer to an `int`, it is a value type with only `const` member functions.  What difference does it make to the user to return a `const FieldPath`?  And why not simplify the code if it makes no difference?

chemelnucfin replied:
To try to convince you, what about `const` member functions but doesn't modify anything in the class?  Would you make the member function `const` or would you say it doesn't matter because it's not modifying the class?

for example, 
`size_t size() {return 0;}`
or 
`size_t size() const {return 0;}`?

I tend to think C++ is in lockdown mode and things should be const unless they have to not be.  
I did make the change in a commit though.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+class FieldPath {
+ public:
+  // Construct FieldPath from a vector of field names
+  FieldPath(const std::vector<std::string>& parts);
```
The rest of the project uses `std::vector<std::string> const&`.  C++ types are read from right to left: a reference to a constant vector of strings.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+
+  // The representation of this FieldPath for ostream
+  friend std::ostream& operator<<(std::ostream& os,
+                                  const FieldPath& field_path);
```
`FieldPath const&` please.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+
+ private:
+  // Ensures string has no invalid characters
+  static void check_invalid_characters(const std::string& string);
```
`std::string const&` please.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+  static std::vector<std::string> split(const std::string& string);
+
+  // The components of this FieldPath
+  std::vector<std::string> parts;
```
s/parts/parts_/

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,73 @@
+  std::vector<std::string> parts;
+};
+}  // namespace firestore
+#endif
```
Needs a blank line and a 

```C++
#endif  // OOGLE_CLOUD_CPP_BIGTABLE_FIRESTORE_GOOGLE_FIRESTORE_FIELD_PATH_H_
```

coryan commented on firestore/tests/field_path_test.cc:
```diff
@@ -0,0 +1,210 @@
+// Copyright 2018 Google Inc.
```
s/Inc./LLC/

Also, we keep the tests in the same directory as the code they test, please move this to `firestore/client`.

coryan commented on firestore/google/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+add_subdirectory(firestore)
```
I think you need to remove these empty `CMakeLists.txt` files.

coryan commented on firestore/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+# add_dependencies(skip-scanbuild-targets gmock firestore_protos)
+
+# All tests get added to this target below.
+add_subdirectory(google)
```
Please remove, there are no files in the `google/` subdirectory.

coryan commented on firestore/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+# All tests get added to this target below.
+add_subdirectory(google)
+add_subdirectory(client)
+add_subdirectory(tests)
```
Please remove, the files in the `tests/` subdirectory should go into `client/`.

coryan commented on firestore/CMakeLists.txt:
```diff
@@ -0,0 +1,129 @@
+
+# All tests get added to this target below.
+add_subdirectory(google)
+add_subdirectory(client)
```
You need to create a CMakeLists.txt file in `client/` that compiles the library.

coryan commented on firestore/client/field_path.cc:
```diff
@@ -0,0 +1,97 @@
+namespace firestore {
+std::regex FieldPath::simple_field_name("[_a-zA-Z][_a-zA-Z0-9]*");
+
+FieldPath::FieldPath(std::vector<std::string> const& parts) : parts_(parts) {
```
Consider receiving `parts` by value so you can use move semantics.

chemelnucfin replied:
If you have time, could you remind me how does this work again?  I do somewhat recall that in C++11 there's no harm in passing by value, but I thought the initializer would automatically move/copy as necessary.  Should I always pass by value?

coryan commented on firestore/client/field_path.cc:
```diff
@@ -0,0 +1,97 @@
+
+FieldPath::FieldPath(std::vector<std::string> const& parts) : parts_(parts) {
+  for (const auto part : parts) {
+    if (part.empty()) {
```
Shouldn't this check for valid characters across the board?

chemelnucfin replied:
invalid characters are only checked in the from_string.  It is a specification detail from firestore.

coryan commented on firestore/client/field_path.cc:
```diff
@@ -0,0 +1,97 @@
+
+const FieldPath FieldPath::from_string(std::string const& string) {
+  check_invalid_characters(string);
+  const auto parts = split(string);
```
If you make `split()` return non-const and the `FieldParts()` constructor take a vector by value then you can write:

```C++
FieldPath FieldPath::from_string(std::string const& string) {
  check_invalid_characters(string);
  return FieldPath(split(string));
}
```

and save yourself a number of copies.

chemelnucfin replied:
Would there be compiler optimizations available due to const?

coryan commented on firestore/client/field_path.cc:
```diff
@@ -0,0 +1,97 @@
+      s += '`' + part + "`.";
+    }
+  }
+  return s.substr(0, s.size() - 1);
```
Consider resizing `s` and returning it to take advantage of the RVO.

chemelnucfin replied:
is this because `s.substr` returns a new string whereas `s.resize` just resizes `s`?

coryan commented on firestore/client/field_path.cc:
```diff
@@ -0,0 +1,97 @@
+  return s.substr(0, s.size() - 1);
+}
+
+bool FieldPath::operator==(FieldPath const& other) const {
```
Does it make sense to define a `operator<` ?

chemelnucfin replied:
I noticed in java that operator < is defined by api_repr, but that would make all unicode characters be between A-Z and a-z.  Maybe someone on the firestore team can confirm?

chemelnucfin replied:
I found out that the java implementation was wrong, and thus followed the node implementation of the operator <.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,74 @@
+  static const FieldPath from_string(std::string const& string);
+
+  // The regex for a simple field name
+  static std::regex simple_field_name;
```
Would the application developers need this?  If not, then consider moving it to the anonymous namespace in the `.cc` file.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,74 @@
+  FieldPath(std::vector<std::string> const& parts);
+
+  // Construct FieldPath from a field path string
+  static const FieldPath from_string(std::string const& string);
```
The naming conventions in Google code (including this project) require functions (and member functions) to use `CamelCase` unless they are a simple accessor.  I do not like them either :disappointed:

chemelnucfin replied:
I know I would be going against the entire Google style guide, but since this is C++ in which the standard library uses underscores, and this is a new project.  Can we reconsider?  I made a commit to change that though.

coryan commented on firestore/client/version_info.h:
```diff
@@ -0,0 +1,22 @@
+#ifndef GOOGLE_CLOUD_CPP_FIRESTORE_CLIENT_VERSION_INFO_H_
+#define GOOGLE_CLOUD_CPP_FIRESTORE_CLIENT_VERSION_INFO_H_
+
+#define FIRESTORE_CLIENT_VERSION_MAJOR 
```
Typically `version_info.h` is generated from a `version_info.h.in` because we wanted to keep the version numbers in just one place.

chemelnucfin replied:
got it.

coryan commented on firestore/client/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+	field_path.cc
+	)
+target_link_libraries(firestore_client PUBLIC
+        firestore_protos)
```
If you do not have a `filestore_protos` library defined we should remove this.

coryan commented on firestore/client/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+        field_path.h
+	field_path_test.cc)
+
+#         testing/chrono_literals.h
```
I know you are developing right now and there is a lot of experimentation needed, but please remember to remove the commented out blocks.

coryan commented on firestore/client/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+#     PRIVATE firestore_common_options
+#)
+
+# List the unit tests, then setup the targets and dependencies.
```
I think you will need this to get the unit test compiled.

coryan commented on firestore/client/CMakeLists.txt:
```diff
@@ -0,0 +1,125 @@
+    add_test(NAME ${target} COMMAND ${target})
+endforeach ()
+
+option(FORCE_SANITIZER_ERRORS
```
You should remove the code from here to line 100.

coryan commented on firestore/CMakeLists.txt:
```diff
@@ -0,0 +1,64 @@
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+
+# Install proto generated files into firestore/google.
+install(DIRECTORY ${CMAKE_BINARY_DIR}/firestore/google/ DESTINATION include/google FILES_MATCHING PATTERN "*.h")
```
You can remove this line I think, since we are not generating any code from the protos (yet).

coryan commented on firestore/CMakeLists.txt:
```diff
@@ -0,0 +1,64 @@
+# The exports can only be installed if all the dependencies are installed.
+# CMake needs to know where the submodules will be installed from,
+install(EXPORT firestore-targets DESTINATION ${CMAKE_INSTALL_DATAROOTDIR}/cmake)
+# endif ()
```
Remove commented out `endif()`

chemelnucfin commented:
I did a `git grep` on `gtest`, and it looks like it's the same as `bigtable` and `storage`.  However, I wasn't able to figure out how to get `gtest` into the `CMakeFile`.

It also looks like visual studio had some problems with C++11?

Again, no rush.  I don't want to interrupt your work on bigtable.
Thanks.

chemelnucfin commented:
@coryan Do you know how I would debug the case [here](https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/jobs/372536712) and  
[here](https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/jobs/372536715)?

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,154 @@
+  bool operator>=(FieldPath const& other) const;
+
+  /**
+   * The representation of this FieldPath *field_path* for ostream *os*.
```
The correct way to format parameters with Doxygen is `@p field_path`.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,154 @@
+
+  /**
+   * The representation of this FieldPath *field_path* for ostream *os*.
+   * @param os.  ostream to write to.
```
I think Doxygen is getting confused by the period.  s/@param os./@param os/.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * The representation of this FieldPath *field_path* for ostream *os*.
+   * @param os.  ostream to write to.
+   * @param field_path.  This const FieldPath.
```
Same thing here.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,154 @@
+
+  /**
+   * Compare if this FieldPath is after another FieldPath *other*.
+   * @param other.  The other const FieldPath to compare to.
```
In general we leave a blank comment line between the short description and the parameter descriptions, please add it here and everywhere else.

coryan commented on firestore/client/field_path.h:
```diff
@@ -0,0 +1,154 @@
+
+ private:
+  /**
+   * Ensures *string* has no invalid characters.
```
Please use `@p string` here and everywhere else.

coryan commented on firestore/client/field_path_test.cc:
```diff
@@ -0,0 +1,188 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "field_path.h"
```
Please use the full path for this include, `#include "firestore/client/field_path.h"`.

chemelnucfin replied:
sorry, must have changed it for testing and then forgot to change it back.

chemelnucfin commented:
Thank you for your patience.

I did try installing the failing test from earlier and wasn't able to run the docker for it after doing `install-linux.sh`
```
>>>TRAVIS_OS_NAME=linux DISTRO=ubuntu DISTRO_VERSION=17.10 CXX=g++ CC=gcc CHECK_STYLE=yes GENERATE_DOCS=yes ./ci/build-linux.sh
No such file or directory
No such file or directory
```
Not too sure what's going on there.

coryan commented:
Congrats!  And thank you!
